webapi/setaltsig.go

```diff
@@ -0,0 +1,137 @@
+	GetRawTransaction(txHash string) (*dcrdtypes.TxRawResult, error)
+}
+
+// setAltSig is the handler for "POST /api/v3/setaltsig".
+func setAltSig(c *gin.Context) {
+
+	const funcName = "setAltSig"
+
+	// Get values which have been added to context by middleware.
+	sigAddress := c.MustGet("SigningAddress").(string)
+	dcrdClient := c.MustGet("DcrdClient").(Node)
+	reqBytes := c.MustGet("RequestBytes").([]byte)
+
+	if cfg.VspClosed {
+		sendError(errVspClosed, c)
+		return
+	}
+
+	var request SetAltSigRequest
+	if err := binding.JSON.BindBody(reqBytes, &request); err != nil {
+		log.Tracef("%s: Bad request (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+
+	altPubAddr, ticketHash := request.AltPubAddress, request.TicketHash
+
+	currentAltSigAddr, err := db.AltSigAddr(ticketHash)
+	if err != nil {
+		log.Errorf("%s: db.AltSigExists (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if currentAltSigAddr != "" {
+		msg := "alternate signature address already exists"
+		log.Errorf("%s: %s (ticketHash=%s)", funcName, msg, ticketHash)
+		sendErrorWithMsg(msg, errBadRequest, c)
+		return
+
+	}
+
+	// Disallow empty addresses.
+	if altPubAddr == "" {
+		const details = "Cannot add new blank alt sig address"
+		log.Tracef("%s: %s (clientIP=%s): %v", funcName, details, c.ClientIP(), err)
+		sendErrorWithMsg(details, errBadRequest, c)
+		return
+	}
+
+	// Fail fast if the pubkey doesn't decode properly.
+	addr, err := stdaddr.DecodeAddressV0(altPubAddr, cfg.NetParams)
+	if err != nil {
+		log.Tracef("%s: Alt pub address cannot be decoded (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+	if _, ok := addr.(*stdaddr.AddressPubKeyHashEcdsaSecp256k1V0); !ok {
+		log.Tracef("%s: Alt pub address is unexpected type (clientIP=%s type=%T): %v", funcName, c.ClientIP(), addr, err)
+		sendErrorWithMsg("wrong type for alternate signing address", errBadRequest, c)
+		return
+	}
+
+	// Get ticket details.
+	rawTicket, err := dcrdClient.GetRawTransaction(ticketHash)
+	if err != nil {
+		log.Errorf("%s: dcrd.GetRawTransaction for ticket failed (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+
+	// Ensure this ticket is eligible to vote at some point in the future.
+	canVote, err := dcrdClient.CanTicketVote(rawTicket, ticketHash, cfg.NetParams)
+	if err != nil {
+		log.Errorf("%s: dcrd.CanTicketVote error (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if !canVote {
+		log.Tracef("%s: unvotable ticket (clientIP=%s, ticketHash=%s)",
+			funcName, c.ClientIP(), ticketHash)
+		sendError(errTicketCannotVote, c)
+		return
+	}
+
+	sigStr := c.GetHeader("VSP-Client-Signature")
+	sig, err := base64.StdEncoding.DecodeString(sigStr)
```

Comment:
Does this need to be decoded?

Reply:
Apparently? It is a string in base64 in the header it seems. Is wrong here? Is working tho.

Reply:
Or you mean just store as string? That's fine...

Reply:
Saving as string.

Reply:
9aebd65